GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs

```diff
@@ -98,140 +96,147 @@ public class ObjAiBase : AttackableUnit, IObjAiBase
                 IsMelee = true;
             }
 
-            AppliedBuffs = new IBuff[256];
-            BuffGameScriptControllers = new List<BuffGameScriptController>();
-            BuffsLock = new object();
+            BuffSlots = new IBuff[256];
+            BuffList = new List<IBuff>();
+            _buffsLock = new object();
             Buffs = new Dictionary<string, IBuff>();
             IsDashing = false;
         }
 
-        public void AddBuffGameScript(string buffNamespace, string buffClass, ISpell ownerSpell, float removeAfter = -1f, bool isUnique = false)
+        public void AddBuff(IBuff b)
         {
-            if (isUnique)
+            lock (_buffsLock)
             {
-                RemoveBuffGameScriptsWithName(buffNamespace, buffClass);
-            }
-
-            var buffController =
-                new BuffGameScriptController(_game, this, buffNamespace, buffClass, ownerSpell, removeAfter);
-            BuffGameScriptControllers.Add(buffController);
-            buffController.ActivateBuff();
-
-            // TODO: should handle the controllers in the class, and don't pass them outside
-        }
-
-        public void RemoveBuffGameScript(BuffGameScriptController buffController)
-        {
-            buffController.DeactivateBuff();
-            BuffGameScriptControllers.Remove(buffController);
-        }
+                if (!Buffs.ContainsKey(b.Name))
+                {
+                    if (HasBuff(b.Name))
```

Comment:
How is this if ever results with true? Please explain.

Reply:
Buffs is the dictionary of parent buffs. HasBuff checks the BuffList, which is a list of all buffs, not just parent buffs. 
The if statement for HasBuff returns true when the unit has a buff of the same name, but no parent buff for buffs of that name.
This if statement is only true for STACKS_AND_OVERLAPS BuffAddType buffs.